Allow WMP in cross_sections.xml, other data updates

paulromano opened the pull request:
This PR makes some data-related updates. Namely, it's now possible to list windowed multipole data files in `cross_sections.xml` where they appear with `type="wmp"`. This means that if you have a `cross_sections.xml` file with WMP data, you don't need to set the `OPENMC_MULTIPOLE_LIBRARY` environment variable. Along the way, I moved the code related to reading `cross_sections.xml` over to C++, which gets rid of the `Library` type and `libraries` / `library_dict` global variables on the Fortran side. There are a few other miscellaneous changes:

- I've added two scripts:
  1. `openmc-convert-lib80x-data` -- This converts the LANL-based [Lib80x](https://nucleardata.lanl.gov/ACE/Production/Lib80x.html) ACE library
  2. `openmc-make-test-data` -- This produces our NNDC-based cross section library that is used for testing. Over time, we've made some custom changes/additions to the data, so it's not sufficient at this point to just download the NNDC ACE files and convert them.
- `openmc_material_set_density` now takes an extra argument for specifying the units of density. We need this for coupling applications where a TH solver might give us updated densities in g/cm3. It would be quite cumbersome to manually convert that to atom/b-cm.
- Combined `string_functions.h` and `string_utils.h`
- Added support for Python's `pathlib` module in `openmc.data`

paulromano commented:
Any willing reviewer for this PR? @openmc-dev/committers

liangjg commented on src/input_xml.F90:
```diff
@@ -2578,20 +2298,38 @@ subroutine read_multipole_data(i_table)
+        else
+          filename = trim(path_multipole) // trim(zero_padded(nuc % Z, 3)) // &
+               trim(zero_padded(nuc % A, 3)) // ".h5"
+        end if
```
Are you planning to remove the 'hacky' way of looking for the wmp library?

liangjg replied:
The new way is definitely clearer and simpler. And I found the new way makes it easier to check if a wmp library exists before reading the point-wise library, this will be helpful for implementation of elastic scattering using multipole data instead of 0K point-wise data. Therefore, I prefer removing the old way. The only downside of to me is the discarding of `OPENMC_MULTIPOLE_LIBRARY`, the users may need to update the `cross_sections.xml` manually to include the wmp library. The travis CI should do this step too.

paulromano commented:
@liangjg I've made a number of updates to this PR:

- I've removed the capability to specify WMP data via the `OPENMC_MULTIPOLE_LIBRARY` environment variable and the `<multipole_library>` element.
- I've updated our test data to include the WMP version 1.1 library.
- While I was updating the data, I decided to put in a fix for #1019 wherein when only level reactions are available for things like `(n,p)`, `openmc.data` will now add the corresponding redundant summed reaction and put it in the HDF5 file. On the Fortran side, if it sees that both the total (e.g. MT=103) and levels (MT=600-649) are present, it will now use the levels when adding up a total (MT=1) cross section. This resulted in some very small changes in test results because before we were using the total when available. As @salcedop pointed out to me, there are a few nuclides in ENDF/B-VII.1 where adding up the level cross sections gives you a different value than summed values in MT=103-107.
- As a result of the previous change, I decided to increment the version of the HDF5 data because of the assumptions OpenMC makes when calculating total cross sections. This ensures that anyone using the develop branch will have to have HDF5 files where each reaction is specifically marked as redundant or not.

liangjg commented:
@paulromano The updates look good to me. Just two more files you may consider updating:
- `docs/source/io_formats/cross_sections.rst`: add `photon` and `wmp` to `library/type`
- `Dockerfile`: remove the `OPENMC_MULTIPOLE_LIBRARY` environment too

paulromano commented:
@liangjg thanks for the suggestions -- just incorporated those changes.

paulromano commented:
@liangjg Any more requests on this PR?

paulromano commented:
@liangjg Just realized you weren't on the committers team -- I sent you an invite so after you accept, you should be able to merge.

liangjg commented:
@paulromano Thanks. Merging!
